File: Java/268-Missing_Number.java
/*Link Of Question : https://leetcode.com/problems/missing-number/
268. Missing Number
Given an array nums containing n distinct numbers in the range [0, n], return the only number in the range that is missing from the array.
*/
//Solution :

class Solution {
    public int missingNumber(int[] arr) {
        int i = 0, n = arr.length;
        int temp = 0, ele = 0;
        
        //This is using Cyclic sort 
        //we can do this by using N*(N+1)/2 Formula, where N natural numbers
        while(i<arr.length)
        {
            ele = arr[i];
            if(ele == n) i++;
            else if(ele == i) i++;
            else
            {
              temp = arr[i];
              arr[i] = arr[ele];
              arr[ele] = temp;
            }
        }
        i = 0;
        while(i<n)
        {
            if(arr[i] != i) return i;
            i++;
        }
        
        return n;
    }
}

/*
Input: nums = [3,0,1]
Output: 2
*/

File: Python/268-missing_number.py
# Link to problem:  https://leetcode.com/problems/missing-number/submissions/

# Problem : Given an array nums containing n distinct numbers in the range [0, n], return the only number in the range that is missing from the array.

#Disclaimer : This might not work in your system as input was handled by leetcode only, you can run this code as it is on the problem link
class Solution:
    def missingNumber(self, nums: List[int]) -> int:
        n =len(nums)
        for i in range(n+1):   #range is [0,n] both inclusive
            if i not in nums:   #return the index whenever it is not in the list, that index will be our answer
                return i
            
'''
Input: nums = [3,0,1]
Output: 2

Input: nums = [0,1]
Output: 2

Time Complexity: O(n)
Space COmplexity: O(1)
'''
